crates/pico/tests/integration_tests.rs

```diff
@@ -0,0 +1,62 @@
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+use calc::{ast::Program, error::Result, eval::eval, lexer::Lexer, parser::Parser};
+use pico::storage::DefaultStorage;
+use pico_core::{database::Database as Db, source::SourceId};
+use pico_macros::{memo, Db, Source};
+
+mod calc;
+
+static COUNTER: AtomicUsize = AtomicUsize::new(0);
+
+#[test]
+fn returns_result() {
+    let mut db = Database::default();
+    let key = "test_expr";
+
+    let mut input = Input {
+        key,
+        value: "2 + 2 * 2".to_string(),
+    };
+    let id = db.set(input);
+    let mut result = evaluate_input(&mut db, id);
+    assert_eq!(result, 6);
+
+    input = Input {
+        key,
+        value: "3 + 3".to_string(),
+    };
+    let id = db.set(input);
+    result = evaluate_input(&mut db, id);
+    assert_eq!(result, 6);
+
+    assert_eq!(COUNTER.load(Ordering::SeqCst), 4);
+}
+
+#[derive(Debug, Default, Db)]
+struct Database {
+    pub storage: DefaultStorage<Self>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Source)]
+struct Input {
+    #[key]
+    pub key: &'static str,
+    pub value: String,
+}
+
+#[memo]
+pub fn parse_ast(db: &mut Database, id: SourceId<Input>) -> Result<Program> {
```

Comment:
Can you add a really basic, easy-to-follow test like:

```
#[memo]
function both_sides(db: &mut Database, left_id: SourceId<Side>, right_id: SourceId<Side>) -> String {
  both_sides_counter.increment();
  let left = calc_side(db, left_id)
  let right = calc_side(db, right_id)
  format!("{} - {}", left, right)
}

function calc_side(&mut Database, id: SourceId<Side>) {
  side_counter_map.get(id).increment();
  let side = db.get(id);
  format!("side text {}", side.text)
}

#[source]
struct Side { text: String }
```

I mean this is a really silly example with no semantic meaning, so maybe it could be something slightly more meaningful, like formatting parts of an address? Idk